### ISSUE TITLE
Upstream Access Log can be incorrect if upstream stream outlives downstream 

### ISSUE THREAD

**Issue by KBaichoo** [envoyproxy/envoy#26413]
Access Log: Can be incorrect if upstream stream outlives downstream 

*Description*:
>What issue is being seen? Describe what should be happening instead of
the bug, for example: Envoy should not crash, the expected value isn't
returned, etc.

As demonstrated by https://github.com/envoyproxy/envoy/pull/26196/files#diff-478cdae3dba52e78ed6d1a2e4e4eb4a1aba3e80178b6c8d14707501fc1fb2ddbR3413-R3498

There are some edge scenarios where the upstream access logs will not reflect the actuality of the upstream stream due to when the access logs are captured -- on downstream stream destruction.

**Comment by KBaichoo** [comment]
cc @yanavlasov  @alyssawilk @RyanTheOptimist

**Comment by alyssawilk** [comment]
yeah IMO we should move upstream logging to upstream filters, which also handles an issue I talked to someone about last week where we don't capture per-upstream-filter header mutations
